test/celeritas/optical/ValidationUtils.hh

```diff
@@ -0,0 +1,186 @@
+{
+  public:
+    /*!
+     * Create a \c real_type \c Span for the supplied collection type \c T.
+     *
+     * If \c T::value_type is \c real_type, then a span to the original data
+     * is returned. Otherwise, it is copied to a local buffer and cast to
+     * \c real_type.
+     */
+    template<class T>
+    Span<real_type const> operator()(T const& xs)
+    {
+        if constexpr (std::is_same_v<real_type,
+                                     std::remove_cv_t<typename T::value_type>>)
+        {
+            return make_span(xs);
+        }
+        else
+        {
+            buffer_.resize(xs.size());
+            for (auto i : range(xs.size()))
+            {
+                buffer_[i] = static_cast<real_type>(xs[i]);
+            }
+            return make_span(buffer_);
+        }
+    }
+
+  private:
+    std::vector<real_type> buffer_;
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Perform consistency checks on grids built in \c Collections.
+ */
+class GridValidator
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Grid = GenericGridRecord;
+    using GridId = OpaqueId<Grid>;
+    using ImportPhysicsTable = std::vector<ImportPhysicsVector>;
+
+    template<class T>
+    using Items = Collection<T, Ownership::value, MemSpace::host>;
+    //!@}
+
+    //! Whether to use soft or exact equivalence
+    enum Softness
+    {
+        Soft,
+        Exact
+    };
+
+  public:
+    // Construct validator for underlying storage
+    GridValidator(Items<real_type>* reals, Items<Grid>* grids);
+
+    // Check the imported data is built under the given grid ID range
+    void check_built_table(ImportPhysicsTable const& table,
+                           ItemRange<Grid> grid_ids,
+                           Softness soft);
+
+    // Check the imported data is built under the given grid ID
+    void check_built_grid(ImportPhysicsVector const& expected,
+                          GridId grid_id,
+                          Softness soft);
+
+    //! Check the imported data is built in the given data range
+    template<class T>
+    void check_built_vector(T const& t,
+                            ItemRange<real_type> const& real_ids,
+                            Softness soft)
+    {
+        this->check_span(convert_real_(t), real_ids, soft);
+    }
+
+    // Construct an MFP builder with the underlying collections
+    MfpBuilder create_mfp_builder();
+
+  private:
+    Items<real_type>* reals_;
+    Items<Grid>* grids_;
+    RealSpanGenerator convert_real_;
+
+    // Check the imported data is built in the given data range
+    void check_span(Span<real_type const> const& t,
+                    ItemRange<real_type> const& real_ids,
+                    Softness soft);
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * A \c GridValidator that stores its own collections.
+ */
+class GridStorage : public GridValidator
```

Comment:
This should be something like `OwningGridValidator`. Its purpose isn't to store; its purpose is validating with local storage.